tests/common/mqtt/test_wrapper.py

```diff
@@ -0,0 +1,232 @@
+    wr = MQTTWrapper("foo", BROKER_INFO)
+
+    assert wr._client_name == "foo"
+    assert wr._last_will is None
+    assert len(wr._dependencies) == 0
+    assert len(wr._dependency_events) == 0
+    assert wr._no_dependency_event is None
+
+    assert len(wr._topic_handlers) == 1
+    assert wr._topic_handlers[Topic(["astoria", "+"])] == wr._dependency_message_handler
+
+    assert wr._client._client_id == "foo"
+
+
+def test_wrapper_init_dependencies() -> None:
+    """Test that the wrapper constructor sets up dependencies."""
+    deps = ["foo", "bar", "bees"]
+    wr = MQTTWrapper("foo", BROKER_INFO, dependencies=deps)
+
+    assert len(deps) == len(wr._dependency_events)
+    assert len(deps) == len(wr._dependencies)
+
+    for d in deps:
+        assert d in wr._dependency_events
+        assert d in wr._dependencies
+
+
+def test_wrapper_init_last_will() -> None:
+    """Test that the wrapper constructor sets up the last will."""
+    lw = ManagerMessage(status="RUNNING")
+    wr = MQTTWrapper("foo", BROKER_INFO, last_will=lw)
+    assert wr._last_will is lw
+
+
+def test_wrapper_init_event() -> None:
+    """Test that the no_dep_event is passed properly."""
+    ev = asyncio.Event()
+    wr = MQTTWrapper("foo", BROKER_INFO, no_dependency_event=ev)
+
+    assert wr._no_dependency_event is ev
+
+
+def test_wrapper_is_connected_at_init() -> None:
+    """Test that the wrapper is not connected to the broker at init."""
+    wr = MQTTWrapper("foo", BROKER_INFO)
+    assert not wr.is_connected
+
+
+def test_wrapper_last_will_message_null() -> None:
+    """Test that the last will message is None when not supplied."""
+    wr = MQTTWrapper("foo", BROKER_INFO)
+    assert wr.last_will_message is None
+
+
+def test_wrapper_last_will_message() -> None:
+    """Test that the wrapper gives a valid last will message."""
+    lw = ManagerMessage(status="RUNNING")
+    wr = MQTTWrapper("foo", BROKER_INFO, last_will=lw)
+
+    message = wr.last_will_message
+    assert isinstance(message, gmqtt.Message)
+    assert message.topic == b"astoria/foo"
+    assert message.qos == 0
+    assert message.retain
+    assert message.payload == lw.json().encode()
+
+
+def test_wrapper_mqtt_prefix() -> None:
+    """Test that the MQTT prefix is as expected."""
+    wr = MQTTWrapper("foo", BROKER_INFO)
+    assert wr.mqtt_prefix == "astoria/foo"
+
+
+def test_subscribe() -> None:
+    """Test that subscribing works as expected."""
+    wr = MQTTWrapper("foo", BROKER_INFO)
+
+    assert len(wr._topic_handlers) == 1
+    assert wr._topic_handlers[Topic(["astoria", "+"])] == wr._dependency_message_handler
+
+    wr.subscribe("bees/+", stub_message_handler)
+    assert len(wr._topic_handlers) == 2
+    assert wr._topic_handlers[Topic(["astoria", "+"])] == wr._dependency_message_handler
+    assert wr._topic_handlers[Topic(["astoria", "bees", "+"])] == stub_message_handler
+
+
+@pytest.mark.filterwarnings("ignore")
```

Comment:
what's this ignoring, ooi?

Reply:
`pytest-asyncio` was giving warnings when parts of `gmqtt` weren't awaiting properly. It was a bit weird tbh.